lib/matplotlib/cbook/__init__.py

```diff
@@ -175,13 +175,20 @@ class CallbackRegistry:
        The default handler prints the exception (with `traceback.print_exc`) if
        an interactive event loop is running; it re-raises the exception if no
        interactive event loop is running.
+
+    signals : list, optional
+        If not None, *signals* is a list of signals that this registry handles:
+        attempting to `process` or to `connect` to a signal not in the list
+        throws a `ValueError`.  The default, None, does not restrict the
+        handled signals.
```

Comment:
This is a *very* useful addition. Not being able to see easily which signals are available on a registry has always been a source of confusion and uncertainty to me.

Is there a reason `signals` is optional (other than backward compatibility)? If not, I propose to deprecate not specifying signals.

Reply:
backend_tools (MEP22) currently relies on being able to dynamically add signals.  I'm not actually convinced that it's a good idea, but that's another discussion to have some other time...

Reply:
Can we have a private `_register_signal()` method for that case? Since `CallbackRegistry` is technically a public API, I'm in favor of changing the signature only once.

Reply:
@tacaswell was also against that (https://github.com/matplotlib/matplotlib/issues/8839#issuecomment-313211155).  I personally don't have a very strong opinion either way, though (well, I think the whole CallbackRegistry API could be revisited, but that's another discussion :-)).

Reply:
I'm not following the "not knowing is a feature" argument. It's comparable to adding attributes at runtime. It can be helpful in very specific cases, but in general it's rather confusing to not know the capabilities of an object.

Anyway, it's not important enough that I would block. If others want to approve as is that's fine for me.

Reply:
I mostly stand by my comment (I do not always agree with myself from 2017!), but I agree that not knowing what keys are supported can be very confusing and I really like the analogy to runtime adding attributes.

I would like to preserve the ability to have a any-to-any registry, I think having a "like known and warn on unknown" behavior is a bad mixed-state with the worst of both worlds.

I am in favor of making `signals` mandatory eventually with the ability to pass `None` to get the naive dispatcher.  

I think adding a (public) API to manage the known signals is also a good idea.